### DISCUSSION TITLE
ch7/ch7-11

### DISCUSSION THREAD

**Discussion by giscus** [gopl-zh/gopl-zh.github.com#179] [General]
# ch7/ch7-11

<The Go Programming Language>中文版

https://gopl-zh.github.io/ch7/ch7-11.html

**Comment by WuHaoYu21** [comment]
这里已经彻底变成机翻了。

**Comment by xubai-owo** [comment]
这纯机翻吧，学go的同时还得分点脑力去理解这绕口的翻译。。。。。
还不如让ai用大白话翻译一遍好理解：

在Go的`os`包中，文件操作可能会返回各种错误。虽然I/O出问题的原因很多，但有三种错误需要特别处理：文件已经存在（比如创建文件时）、文件找不到（比如读取时）、权限不足。`os`包提供了三个函数专门判断这些错误：

```go
package os

func IsExist(err error) bool     // 是不是「文件已存在」？
func IsNotExist(err error) bool  // 是不是「文件不存在」？
func IsPermission(err error) bool // 是不是「没权限」？
```

不要用土办法判断错误

你可能想直接看错误信息里有没有特定文字，比如：

```go
func IsNotExist(err error) bool {
    return strings.Contains(err.Error(), "file does not exist")
}
```

但这种方法不靠谱！因为不同操作系统报错的信息可能不一样，比如在Windows和Linux上同一个错误显示的文字可能不同。虽然测试的时候用字符串匹配可能方便，但正式代码里千万别这么干。

正确的做法：看错误的结构

`os`包定义了一个`PathError`结构体，专门记录文件路径相关的错误（比如打开或删除文件失败），还有`LinkError`记录涉及两个路径的操作错误（比如重命名）。`PathError`长这样：

```go
type PathError struct {
    Op   string    // 操作类型，比如"open"、"delete"
    Path string    // 文件路径
    Err  error     // 具体的错误原因
}

// 拼成错误信息：例如 "open /tmp/foo: permission denied"
func (e *PathError) Error() string { ... }
```

关键点：虽然显示的错误信息是一个字符串，但`PathError`内部保存了错误的结构。你可以像查户口一样，把错误拆开来看具体原因。

例如这段代码：

```go
_, err := os.Open("/不存在的文件")
fmt.Println(err) // 输出：open /不存在的文件: no such file or directory
fmt.Printf("%#v\n", err)
// 输出：&os.PathError{Op:"open", Path:"/不存在的文件", Err:0x2}
```

判断错误的具体类型

比如`IsNotExist`函数的原理是这样的：

1. 先检查错误是不是`PathError`类型
2. 如果是，就看看它内部的`Err`字段是不是系统级的“文件不存在”错误（比如`syscall.ENOENT`），或者`os`包自己定义的`ErrNotExist`

代码实现：

```go
func IsNotExist(err error) bool {
    // 如果是PathError，就取出内部的错误原因
    if pe, ok := err.(*PathError); ok {
        err = pe.Err
    }
    // 判断是否是系统级错误或自定义错误
    return err == syscall.ENOENT || err == ErrNotExist
}
```

使用时很简单：

```go
_, err := os.Open("/不存在的文件")
fmt.Println(os.IsNotExist(err)) // 输出true
```

重要提醒：  
如果你用`fmt.Errorf`这类函数把错误包装成新字符串，原本的`PathError`结构就丢了。所以一定要在出错的地方立刻判断错误类型，不要把错误传来传去之后才处理。
